Consistent types for vector inputs/outputs in base layer

dstansby opened the pull request:
# Fixes/Closes

<!-- In general, PRs should fix an existing issue on the repo. -->
<!-- Please link to that issue here as "Closes #(issue-number)". -->
Closes #

# Description
As part of an ongoing effort to clean up type hints in layers, this PR adds consistent typing for vector input/ouputs in the base layer class. The choices I made were:

- In public API, be as permissible as possible and use `npt.ArrayLike` for typing, and then `np.asarray` to coerce everything into an array as soon as possible.
- In private API, for simplicity and performance use `np.ndarray` everywhere.
- Use `List[int]` for `dims_displayed`, as this allows `int` to be specified, and no arithmetic needs to be done on this variable.

This fixes lots of existing typing errors in `napari.layers.base`, but not all of them. I thought this change was self-contained enough to go in its own PR.

# References
<!-- What resources, documentation, and guides were used in the creation of this PR? -->

## Type of change
<!-- Please delete options that are not relevant. -->
- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How has this been tested?
<!-- Please describe the tests that you ran to verify your changes. -->
- [x] all tests pass with these changes

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

dstansby committed:
Consistent types for vector inputs/outputs in base layer

dstansby committed:
Fix new errors

melonora committed:
Merge branch 'main' into vector-types

melonora commented on napari/layers/base/base.py:
```diff
@@ -1341,7 +1345,7 @@ def refresh(self, event=None):
             self._set_highlight(force=True)
 
-    def world_to_data(self, position):
+    def world_to_data(self, position: npt.ArrayLike) -> np.ndarray:
```
```suggestion
    def world_to_data(self, position: npt.ArrayLike) -> npt.NDArray:
```

melonora commented on napari/layers/base/base.py:
```diff
@@ -1386,8 +1391,8 @@ def data_to_world(self, position):
-        self, position: np.ndarray, dims_displayed: np.ndarray
-    ) -> tuple:
+        self, position: np.ndarray, dims_displayed: List[int]
+    ) -> np.ndarray:
```
```suggestion
    ) -> npt.NDArray:
```

melonora commented on napari/layers/base/base.py:
```diff
@@ -1419,7 +1424,7 @@ def _data_to_world(self) -> Affine:
         return self._transforms[1:3].simplified
 
-    def _world_to_data_ray(self, vector) -> tuple:
+    def _world_to_data_ray(self, vector: npt.ArrayLike) -> np.ndarray:
```
```suggestion
    def _world_to_data_ray(self, vector: npt.ArrayLike) -> npt.NDArray:
```

melonora commented on napari/layers/base/base.py:
```diff
@@ -1437,33 +1442,33 @@ def _world_to_data_ray(self, vector) -> tuple:
 
     def _world_to_displayed_data_ray(
-        self, vector_world, dims_displayed
+        self, vector_world: np.ndarray, dims_displayed: List[int]
```
```suggestion
        self, vector_world: npt.NDArray, dims_displayed: List[int]
```

melonora commented on napari/layers/base/base.py:
```diff
@@ -1437,33 +1442,33 @@ def _world_to_data_ray(self, vector) -> tuple:
     def _world_to_layer_dims(
-        self, *, world_dims: List[int], ndim_world: int
-    ) -> List[int]:
+        self, *, world_dims: np.ndarray, ndim_world: int
```
```suggestion
        self, *, world_dims: npt.NDArray, ndim_world: int
```

melonora commented on napari/layers/base/base.py:
```diff
@@ -1477,28 +1482,32 @@ def _world_to_layer_dims(
+        return order[order >= offset] - offset
 
-    def _display_bounding_box(self, dims_displayed: np.ndarray):
+    def _display_bounding_box(self, dims_displayed: List[int]) -> np.ndarray:
```
```suggestion
    def _display_bounding_box(self, dims_displayed: List[int]) -> npt.NDArray:
```

melonora commented on napari/layers/base/base.py:
```diff
@@ -1477,28 +1482,32 @@ def _world_to_layer_dims(
-    def _display_bounding_box_augmented(self, dims_displayed: np.ndarray):
+    def _display_bounding_box_augmented(
+        self, dims_displayed: List[int]
+    ) -> np.ndarray:
```
```suggestion
    ) -> npt.NDArray:
```

melonora commented on napari/layers/base/base.py:
```diff
@@ -1591,17 +1602,17 @@ def get_ray_intersections(
         return start_point, end_point
 
-    def _get_offset_data_position(self, position: List[float]) -> List[float]:
+    def _get_offset_data_position(self, position: np.ndarray) -> np.ndarray:
```
```suggestion
    def _get_offset_data_position(self, position: npt.NDArray) -> npt.NDArray:
```

melonora commented on napari/layers/base/base.py:
```diff
@@ -1591,17 +1602,17 @@ def get_ray_intersections(
     def _get_ray_intersections(
         self,
-        position: List[float],
+        position: np.ndarray,
```
```suggestion
        position: npt.NDArray,
```

melonora commented on napari/layers/base/base.py:
```diff
@@ -1591,17 +1602,17 @@ def get_ray_intersections(
+        position: np.ndarray,
         view_direction: np.ndarray,
         dims_displayed: List[int],
+        bounding_box: np.ndarray,
```
```suggestion
        bounding_box: npt.NDArray,
```

melonora commented on napari/layers/base/base.py:
```diff
@@ -1649,8 +1660,9 @@ def _get_ray_intersections(
 
-            view_dir = np.asarray(view_direction)[dims_displayed]
-            click_pos_data = np.asarray(position)[dims_displayed]
+            print(type(position))
```
```suggestion
```

melonora commented on napari/layers/base/base.py:
```diff
@@ -1847,27 +1860,32 @@ def get_status(
     def _get_tooltip_text(
         self,
-        position,
+        position: np.ndarray,
```
```suggestion
        position: npt.NDArray,
```

melonora commented on napari/layers/image/image.py:
```diff
@@ -886,15 +886,15 @@ def _get_value(self, position):
         return value
 
-    def _get_offset_data_position(self, position: List[float]) -> List[float]:
+    def _get_offset_data_position(self, position: np.ndarray) -> np.ndarray:
```
```suggestion
    def _get_offset_data_position(self, position: npt.NDArray) -> npt.NDArray:
```

melonora reviewed (changes requested):
I think in general `npt.NDArray` can be used to type here. There could also be an import of `TYPE_CHECKING` and then import `npt`. All minor stuff though:)

Czaki committed:
Update deprecation information (#5984)

# Description
<!-- What does this pull request (PR) do? Why is it necessary? -->
<!-- Tell us about your new feature, improvement, or fix! -->
<!-- If your change includes user interface changes, please add an
image, or an animation "An image is worth a thousand words!" -->
<!-- You can use https://www.cockos.com/licecap/ or similar to create
animations -->

Use proper version in deprecation information 

# References
<!-- What resources, documentation, and guides were used in the creation
of this PR? -->

## Type of change
<!-- Please delete options that are not relevant. -->
- [x] Maintenance (changes required to run napari, tests, & CI smoothly)

# How has this been tested?
<!-- Please describe the tests that you ran to verify your changes. -->
- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt
backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] If I included new strings, I have used `trans.` to make them
localizable.
For more information see our [translations
guide](https://napari.org/developers/translations.html).

---------

Co-authored-by: Wouter-Michiel Vierdag <w-mv@hotmail.com>

dstansby committed:
Use numpy.typing for ndarray hints

Co-authored-by: Wouter-Michiel Vierdag <w-mv@hotmail.com>

dstansby committed:
Guard np.typing in TYPE_CHECKING

melonora committed:
Merge branch 'main' into vector-types

dstansby committed:
Add missing typing import

Czaki commented on napari/layers/base/base.py:
```diff
@@ -1172,11 +1185,11 @@ def _get_value(self, position):
     def get_value(
         self,
-        position: Tuple[float, ...],
+        position: npt.ArrayLike,
```
I have mixed feelings. Previous annotations inform that it is single-dimensional. 
```suggestion
        position: Sequence[float],
```

melonora replied:
hmm I still think npt.ArrayLike fits better as it is described as any `Sequence` that can be converted an array of one `dtype`. But I do understand your point of losing information. I do think it is clear from the argument name though and the function definition.

Czaki replied:
Yes. but for `Sequence[float]` valid is `(1, 1)` 

For `npt.ArrayLike` valid is also `[[1, 1],[1,1]]`

jni replied:
Yeah I agree with @Czaki here — until we can use [variadic generics](https://peps.python.org/pep-0646/) with npt.ArrayLike, `Sequence[float]` is a better fit imho.

jni replied:
(And I *think* that means, until Py3.11 is our minimum supported version, but I don't know whether PEP-646 was backported to earlier versions, or whether NumPy has implemented it yet.)

melonora replied:
ok you convinced me @Czaki. @dstansby can you implement the change please?:)

dstansby replied:
I started making this change, but `Sequence[float]` doesn't alow for something of type `np.ndarray` to be passed in, which we definitely want to allow.

I do agree that where possible we want to specify the number of dimensions, but in the case of positions I decided to go with `ArrayLike` because we want to always cast the input to an array to do array operations (e.g. multiplication, division, norming) on the position.

So for the purpose of keeping the typing simple, I'd advocate for staying with `ArrayLike` instead of `Sequence` - thoughts?

melonora replied:
The only other option I would say is a `Union` of `ArrayLike` and `Sequence`. Or to add the docstring in which it is stated that single dimension is expected and raising a ValueError if position is not passed as a 1-dimensional sequence.

melonora replied:
@jni @Czaki Personally I would say the latter option keeping `ArrayLike` as type hint, but raising `ValueError` if not 1-dimensional is the way to go. What do you think?

jni replied:
Wow I'm quite devastated that ndarray doesn't pass typing as a Sequence[float], but I guess it makes sense. I'm not opposed to that approach then. Union is too clunky.

dstansby committed:
Merge branch 'main' into vector-types

jni commented on napari/layers/base/base.py:
```diff
@@ -52,6 +61,10 @@
 from napari.utils.translations import trans
 
+if TYPE_CHECKING:
+    import numpy.typing as npt
```
Just as a final style issue, and this is totally up for discussion and not a requirement, but to consider: as an old-school Pythonista, I find that typing can get really convoluted very quickly. I think "English readability" is very important, and `npt.` is quite clunky. So what do people think about `from numpy.typing import ArrayLike` and then just using `ArrayLike`?  And you could add a comment next to it `# must be 1D`.

melonora replied:
I would say this would have to be discussed, perhaps open an issue. We use `npt` already like this in multiple locations, for example `test_transforms.py` and `_shapes_utils.py` so if we change the convention it needs to be changed in locations outside the intended scope of this PR.

Personally I am neutral about it though I understand the point of brevity

dstansby replied:
This is because `npt` follows conventions in the `numpy` docs, e.g. at https://numpy.org/doc/stable/reference/typing.html#number-precision.

melonora committed:
Merge branch 'main' into vector-types
